consensus_test.go

```diff
@@ -208,6 +130,96 @@ func TestTransition_AcceptState_Validator_VerifyFails(t *testing.T) {
 	})
 }
 
+func TestTransition_AcceptState_Proposer_FailedBuildProposal(t *testing.T) {
+	buildProposalFunc := func() (*Proposal, error) {
+		return nil, errors.New("failed to build a proposal")
+	}
+
+	validatorIds := []string{"A", "B", "C"}
+	backend := newMockBackend(validatorIds, nil, buildProposalFunc, nil, nil)
+
+	m := newMockPbft(t, validatorIds, "A", backend)
+	m.state.view = ViewMsg(1, 0)
+	m.setState(AcceptState)
+
+	var buf bytes.Buffer
+	m.logger.SetOutput(&buf)
+	defer func() {
+		m.logger.SetOutput(getDefaultLoggerOutput())
+	}()
+
+	// Prepare messages
+	m.emitMsg(&MessageReq{
+		From: "A",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+	m.emitMsg(&MessageReq{
+		From: "B",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+	m.emitMsg(&MessageReq{
+		From: "C",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+
+	go func() {
+		for {
+			if m.GetState() == RoundChangeState {
+				m.cancelFn()
+				return
+			}
+		}
+	}()
+
+	m.Run(m.ctx)
+	m.logger.Println(buf.String())
+
+	assert.Contains(t, buf.String(), "[ERROR] failed to build proposal", "build proposal did not fail")
+	assert.True(t, m.IsState(RoundChangeState))
+}
+
+func TestTransition_AcceptState_Proposer_Cancellation(t *testing.T) {
+	// We are in the AcceptState, we are the proposer.
+	// Cancellation occurs, so we are checking whether we are still in AcceptState.
```

Comment:
What does cancellation occurs means?

Reply:
Generally speaking, intention was to cover all the branches in the code base with unit tests, although this maybe isn't real use case.
This test specifically covers `p.ctx.Done()` channel from the snippet bellow:

```go
func (p *Pbft) runAcceptState(ctx context.Context) {
...
if isProposer {
...
   select {
      case <-time.After(delay):
      case <-p.ctx.Done():
         return
   }
...
```
If you don't see need for such test, I can remove it.